classes/auth.php

```diff
@@ -925,9 +925,12 @@ public function update_user_profile_fields(&$user, $attributes, $newuser = false
                                         continue;
                                     }
                                 }
-
-                                // Custom profile fields have the prefix profile_field_ and will be saved as profile field data.
-                                $user->$field = $attributes[$attr][0];
+                                if ($field == 'username') {
+                                    $user->$field = strtolower($attributes[$attr][0]);
```

Comment:
do we have a test that covers that scenario?

Reply:
yeah - there are some tests that are passing, but there's an edge case here that I don't think we have a test for - I'll try to come back and add one in.

I'm also looking closely at this block:
https://github.com/catalyst/moodle-auth_saml2/blob/MOODLE_39_STABLE/classes/auth.php#L920-L927

if the external idp value is matching on something like idnumber - it should be possible to update the moodle username IMO.

Reply:
Yes agree. It feels like this code needs some refactoring to include all these scenarios and still be understandable.

Reply:
So basically only matching fields should be locked from updating. This will require to add more logic for checking things + tests.

Reply:
in the meantime - I can see a test I can add for this specific case - just running it through now so I can get this fix in and we can come back to the fact that username can't be updated later.

Reply:
Lets' create an issue to come back to it later. Otherwise LGTM.